sycl/source/detail/program_manager/program_manager.cpp

```diff
@@ -2178,24 +2178,52 @@ std::pair<RT::PiKernel, std::mutex *> ProgramManager::getOrCreateKernel(
                         &(BuildResult->MBuildResultMutex));
 }
 
-bool doesDevSupportImgAspects(const device &Dev,
-                              const RTDeviceBinaryImage &Img) {
-  const RTDeviceBinaryImage::PropertyRange &PropRange =
-      Img.getDeviceRequirements();
-  RTDeviceBinaryImage::PropertyRange::ConstIterator PropIt = std::find_if(
-      PropRange.begin(), PropRange.end(),
-      [](RTDeviceBinaryImage::PropertyRange::ConstIterator &&Prop) {
-        using namespace std::literals;
-        return (*Prop)->Name == "aspects"sv;
-      });
-  if (PropIt == PropRange.end())
+bool doesDevSupportDeviceRequirements(const device &Dev,
+                                      const RTDeviceBinaryImage &Img) {
+  auto getPropIt = [&Img](const std::string &PropName) {
+    const RTDeviceBinaryImage::PropertyRange &PropRange =
+        Img.getDeviceRequirements();
+    RTDeviceBinaryImage::PropertyRange::ConstIterator PropIt = std::find_if(
+        PropRange.begin(), PropRange.end(),
+        [&PropName](RTDeviceBinaryImage::PropertyRange::ConstIterator &&Prop) {
+          return (*Prop)->Name == PropName;
+        });
+    return (PropIt == PropRange.end())
+               ? std::nullopt
+               : std::optional<
+                     RTDeviceBinaryImage::PropertyRange::ConstIterator>{PropIt};
+  };
+
+  auto AspectsPropIt = getPropIt("aspects");
+  auto ReqdWGSizePropIt = getPropIt("reqd_work_group_size");
+
+  if (!AspectsPropIt && !ReqdWGSizePropIt)
     return true;
-  ByteArray Aspects = DeviceBinaryProperty(*PropIt).asByteArray();
-  // Drop 8 bytes describing the size of the byte array.
-  Aspects.dropBytes(8);
-  while (!Aspects.empty()) {
-    aspect Aspect = Aspects.consume<aspect>();
-    if (!Dev.has(Aspect))
+
+  // Checking if device supports defined aspects
+  if (AspectsPropIt) {
+    ByteArray Aspects =
+        DeviceBinaryProperty(*(AspectsPropIt.value())).asByteArray();
+    // Drop 8 bytes describing the size of the byte array.
+    Aspects.dropBytes(8);
+    while (!Aspects.empty()) {
+      aspect Aspect = Aspects.consume<aspect>();
+      if (!Dev.has(Aspect))
+        return false;
+    }
+  }
+
+  // Checking if device supports defined required work group size
+  if (ReqdWGSizePropIt) {
+    ByteArray ReqdWGSize =
+        DeviceBinaryProperty(*(ReqdWGSizePropIt.value())).asByteArray();
+    // Drop 8 bytes describing the size of the byte array.
+    ReqdWGSize.dropBytes(8);
+    int ReqdWGSizeAllDimsTotal = 1;
+    while (!ReqdWGSize.empty())
+      ReqdWGSizeAllDimsTotal *= ReqdWGSize.consume<int>();
+    if (static_cast<long unsigned int>(ReqdWGSizeAllDimsTotal) >
+        Dev.get_info<info::device::max_work_group_size>())
```

Comment:
We should also be checking against `max_work_item_sizes`. Keep in mind that `ReqdWGSize` is in reverse order compared to `max_work_item_sizes` and is dimensionality-dependent so you need to compare based on that too.

Reply:
Regarding, e.g., 3-dims, do you mean that if max_work_item_sizes<3>={A, B, C} and reqd_work_group_size is {D, E, F}, we shouldn't check if A*B*C > D*E*F, instead, we should check if A > F, B > E and C > D?

Reply:
Exactly! So the requirements would be:

| `max_work_item_sizes` | `reqd_work_group_size` | Requirement |
|---|---|---|
| `{A}` | `{D}` | `A >= D` |
| `{A, B}` | `{D, E}` | `A >= E && B >= D`  |
| `{A, B, C}` | `{D, E, F}` | `A >= F && B >= E && C >= D` |